Add Currencies library and get and parse rate dynamics

suhanoves opened the pull request:
Игорь, привет! Меня зовут Евгений, я начинающий разработчик и это мой первый опыт, когда я пытаюсь законтрибьютить в опенсорс, поэтому если я что-то делаю не то, ты мне скажи, я буду исправляться.

Теперь к делу: я пользуюсь твоим модулем, но увидел, что на сайте Центробанка есть ещё интересные возможности, которые не покрыты pycbrf, а именно:

1. Библиотека всех иностранных валют с русскими и английскими названиями и кодами
2. Возможность получить по определённой валюте историю курсов на огромный период

Это привело меня к доработке твоего проекта, но его переосмысление немного ломает существующую совместимость. Сейчас поясню.

На текущий момент у тебя сведения о валюте существуют как аттрибуты экземпляра курса. Моя идея состоит в следующем:

1. Создаём библиотеку всех валют, вытягивая и распарсивая с сайте XML. Делаем данную библиотеку Singleton, чтобы она и все её изменения были доступны из любого экзепляра ставок. Это нужно потому, что в библиотеке валют удалены сильно устаревшие или заменённые валюты, которые иногда вылезают при запросе мультивальтного курса.
Дополнительным плюсом является то, что валюта хранит в себе английское название и русское в именительном падеже.
2. Из экземпляра ExchangeRate убираем аттрибуты, привязанные к валюте. Остаётся валюта и параметры курса: дата, стоимость, номинал (с номиналом выяснились тоже интересные особенности, но об этом потом).
3. Экземпляр ExchangeRates делаем универсальным, он теперь может запрашивать и парсить как мультивалютный, так и мультивременной курсы в зависимости от запроса. Если валюта не задана, то значит мы хотим получить мультивалютный курс на дату. Если задана валюта - значит мы хотим получить мультивременной курс для определённой валюты.

Что мной сделано:
1. Библиотека в виде синглтона через мастеркласс. Хранение валют организовано в виде словаря, где для каждой валюты создаётся три ключа: id, num и rate. Я пытался организовать и через словарь + список (аналогично организацию словаря в современном питоне), но оказалось, что так потребление ресурсов даже выше.
Подобная структура позволяет нам за O(1) находить любую существующую валюту по любому определяющему параметру, при этом независимо от того, как она передана: 'usd' 'USD' 'R01235' 'rR01235' '840' 840 c ведущим нулём и без. Валют всего 150, поэтому потребление ресурсов считаю незначительным.
2.  Класс BetaExchangeRate - аналог твоего ExchangeRate, но чуть расширенный. 
_Приставка Beta добавлена потому, что не на 100% сохраняется старый интерфейс, и не все тесты будут автоматически пройдены._
Я считаю неправильным хренение параметров, относящиеся к одному курсу в библиотеке курсов. Поэтому BetaExchangeRate имеет дополнительные аттрибуты date_requested, date_received, dates_match, которые ранее у тебя относились к ExchangeRates.
Для сохранения небольшой совместимости со старым интерфейсом созданы @property методы, которые получают данные из валюты курса, поэтому параметры валюты можно получить сразу обращаясь к экземпляру курса.
3. Класс BetaExchangeRates - аналог твоего ExchangeRates, который теперь умеет парсить в двух вариантах:
  * Если не задана валюта, мы ищем мультивальтный курс на сегодня (если не задана дата) или на заданную дату
  * Если валюта задана мы ищем курс валюты на сегодня (если не задана дата), на другую дату(если дата одна), или на курсы на диапазон дат, если заданы обе даты и валюта.
  * Доступ к экземпляру каждой ExchangeRate осуществляется либо по дате (если курс мультивременной), либо по любому параметру валюты, если курс мультивалютный. Тип BetaExchangeRates определяется по аттрибуту is_multicurrency
4. Добавлены кастомные исключения если не найдена валюта или курс, поэтому не будут проходить твои тесты, где при пустом значении у тебя возвращается None. Мне кажется правильным возбуждать исключения, как в словаре.

Во время парсинга обнаружились интересные моменты:
1. Иногда в мультивальтном курсе могут появиться валюты, которые уже не существуют в библиотеке. Так случилось со старыми белорусскими зайчиками. В этом случае мы добавляем неизвестную валюту в библиотеку и всё работает отлично (разве что оба имени будут на английском), но можно добавить ещё один запрос.
2. Номиналы валют в библиотеке и в курсах могут иногда отличатся. Написал в цетробанк, чтобы допилили библиотеку, но слабо в это верю. Поэтому для всех расчётов номиналы берутся из курса, а не валюты.
3. Обнаружил, что в библиотеке всегда валюта называется в именительном падеже,а в курсах может быть ещё и в родительном и с чем это связано непонятно. Поэтому имена берутся всегда из самой валюты.

Разумеется на всё написана подробная документация, везде расставлены аннотации типов и написано большое количество тестов.

Меня сильно смущает то, что старый интерфейс не на 100% совместим, поэтому жду от тебя твоей реакции и готов править код.
P.S.: Если коммитов слишком много, могу слить их в один.

suhanoves committed:
Add classes Currency and CurrenciesLib

suhanoves committed:
Add tests for Currency and CurrenciesLib

suhanoves committed:
Add __hash__ and __eq__ to Currency

suhanoves committed:
Add and edit docstring, annotations for Currency and CurrenciesLib

suhanoves committed:
Minor fixes in CurrenciesLib methods

suhanoves committed:
Some of the identical methods have been moved to the mixin

suhanoves committed:
Add custom Exceptions

suhanoves committed:
Add 'add' and 'pack' methods, fix docs, minor refactoring

suhanoves committed:
Change datetime.datetime to datetime.date in FormatMixin

suhanoves committed:
Add tests for CurrenciesLib, wrong args and multicurrencies ExchangeRates

suhanoves committed:
Add BetaExchangeRates to __init__.py

suhanoves committed:
Add tests for multidates ExchangeRates

suhanoves committed:
Add BetaExchangeRate and BetaExchangeRates classes, refactoring

suhanoves committed:
Add Exception message tests

suhanoves committed:
Edit docs, type annotations and tests. Minor refactoring

idlesign commented:
Привет.
Спасибо. После беглого осмотра комментарии общего толка:

1. Обратная совместимость, несомненно важна: люди частью просто не перейдут на новую абстракцию, а поддерживать две — сомнительное удовольствие. Поэтому, пожалуй, первое, что следует сделать — это вписать новые плюшки в существующие примитивы:
 * в `ExchangeRate` для самодостаточности действительно просится дата курса (только не запрошенная, а реальная);
 * туда же добавить атрибут `currency`, в котором будет расширенная информация о валюте;
 * а  `id, name, code, num` пусть обращаются к атрибутам этой информации;
 * динамика тоже может быть полезна, но для неё лучше завести `ExchangeRatesDynamics`
2. Нужно подумать над целесообразностью внесения информации по наименованиям и кодам валют, ввиду существования ОКВ, доступного в частности из  https://github.com/idlesign/ruopenrefs (данные там должны быть более полные, но минус в необходимости получения ключа для доступа к данным). Вообще штуки типа ОКВ и ОКСМ относительно редко меняются и небольшие, как вариант следует подумать над тем, чтобы их зашить, а не бегать в сеть.

suhanoves commented:
1. Глянул ОКВ, вижу главный минус - это отсутствие кодов самого центробанка, а именно он нужен для запроса динамики ставок. Предлагаю оставить библиотеку, т.к.:
    * коды центробанка всё равно нужны
    * запрашивается она всего лишь один раз при инициализации любого из классов, а дальше работает один экземпляр
    * для её обновления с сохранением добавленных валют уже реализован метод .update
    * если появится более интересный вариант, реализацию класса можно поменять, а интерфейс останется неизменным
2. Что делать с исключениями? Для унификации так же возвращать None при остутствии данных или в ExchangeRatesDynamics поднимать исключения?
3. Да, хотел спросить и забыл, объекты даты почему были выбраны datetime, а не date? Ведь центробанк не отдаёт время, и курс актуален только на день. В ExchangeRatesDynamics в качестве ключей доступа к ExchangeRate будут выступать объекты даты. Что использовать date или datetime?

suhanoves committed:
Added: Currency, CurrenciesLib, ExchangeRateDynamics. Rewritten: ExchangeRate, ExchangeRates

suhanoves commented:
Игорь, глянь на текущую версию. Она полностью совместима с публичными интерфейсами твоей существующей версии.
Все тесты, что были у тебя проходят на 100% (банки проходят на 66%, один пропускается, но так было и до этого, туда не лез)
Ченжлог:
1. Добавлен класс Currency, представляющий собой валюту с её параметрами.
    * Хешируется для того, чтобы выступать ключём словаря
2. Добавлен класс CurrenciesLib, представляющий собой синглтон и загружающий список валют с сайта ЦБ РФ.
    * Имеет возможность ручного добавления валюты CurrenciesLib.add(), если её не было в библиотеке, но появилась в курсах.
    * Имеет возможность обновления библиотеки без затирания вручную добавленных валют CurrenciesLib.update
    * Представляет собой словарь с ключами из основных аттрибутов валюты (id, num, code), значение выступает экземпляр валюты.
    * Получение валюты по любому её ключевому параметру в любом регистре, str или num для num_code валюты
3. Переписан класс ExchangeRate:
    * Добавлен экземпляр валюты внутрь. Для обратной совместимости созданы property свойства для получения сведений валюты по аттрибутами самого ExchangeRate
4. Переписан класс ExchangeRates:
    * Переписана функциональность под новый ExchangeRate. Теперь аттрбут rates представляет собой словарь, ключём которого выступает валюта, что позволяет искать входящие ExchangeRate по любому ключевому параметру валюты.
    * В __getitem__ возбуждаются кастомные исключения, но ловятся для обратной совместимости со старыми версиями. Возвращают None.
    * Если при получении курсов возникает валюта, отсутствующая в Библиотеке, она добавляется в Библиотеку.
    * Добавлен метод __len__
5. Добавлен класс ExchangeRateDynamics, представляет собой ExchangeRates для одной валюты с разными датами.
    * Аттрибут rates представляет собой словарь с ключём datetimes (сделан для унификации с ExchangeRate), значения ExchangeRate.
    * Способен искать как без аттрибутов даты(на сегодня), с одной датой (на дату), с двумя датами (диапазон)
    * Задан метод __getitem__, может принимать дату в строке, в виде объектов date или datetime
    * Добавлен метод __len__
    * В отличие от ExchangeRate **при неверных параметрах возбуждает исключения, а не возвращает None**, т.к. здесь обратная совместимость не требуется.
 6. Добавлены кастомные классы исключений WrongArguments, CurrencyNotExists, ExchangeRateNotExists
 7. Добавлен миксин FormatMixin для общих форматирований аттрибутов
 8. Добавлен миксин SingletonMeta - метакласс для создания синглтона 
 9. Добавлена и актуализирована документация для всех классов и методов
 10. Добавлены дополнительные тесты для всех классов

idlesign commented:
Спасибо. Постараюсь на неделе поглядеть реализацию.

idlesign commented on pycbrf/__init__.py:
```diff
@@ -1,9 +1,8 @@
 from .banks import Banks
-from .rates import ExchangeRates
-
+from .rates import Currency, CurrenciesLib, ExchangeRates, ExchangeRateDynamics
```
Давай сократим до Currencies, постфикс Lib дополнительной нагрузки не несёт.

idlesign commented on pycbrf/banks.py:
```diff
@@ -2,7 +2,7 @@
 from io import BytesIO
 from logging import getLogger
-from typing import NamedTuple, List, Union, Optional, Set
+from typing import NamedTuple, List, Union, Optional
```
Поскольку в этом запросе на слияние множество коммитов, мы это дело уплотним до одного (squash).
И тогда данное изменение с форматированием в модуле, не касающемся курсов, будет «шуметь» как не касающееся к делу.
Тут два пути: либо убрать изменения, не касающееся непосредственно задачи вовсе, либо изменения по задаче оформить единым коммитом, а изменения не связанные — вторым. Тогда можно будет обойтись без уплотнения вовсе.

idlesign commented on pycbrf/exceptions.py:
```diff
@@ -1,3 +1,23 @@
+        self.message = message
+
+
+class CurrencyNotExists(PycbrfException, KeyError):
```
NotExists лучше заменить на NotFound или Missing, ибо «не существует» и  «не найдено»/«отсутствует» имеют разные смысловые оттенки.

idlesign commented on pycbrf/utils.py:
```diff
@@ -23,3 +25,38 @@ def _get_response(cls, url: str, **kwargs) -> requests.Response:
+    """Mixin for various argument formatting"""
+
+    @staticmethod
+    def _format_num_code(num: Union[int, str]) -> str:
```
Кажется `f'{val}'.zfill(3)` сделает то же шустрее.

suhanoves replied:
не знал о таком методе даже, спасибо!

idlesign commented on pycbrf/utils.py:
```diff
@@ -1,19 +1,21 @@
+import datetime as dt
```
Если из модуля используется только datetime, то можно его сразу и импортировать, чтобы в коде каждый раз не разрешался через точку атрибут модуля.

suhanoves replied:
В коде встречаются переменные date, а dt.date используется в аннотациях типов.
Если убирать dt, то надо переименовывать переменные, либо оставляем импорт as dt.
Как сделать?

idlesign replied:
Если я правильно помню, то изначально  в коде date не должно было быть, вроде везде on_date.
Если где-то в новом коде есть, давай переименуем.

suhanoves replied:
В ExchangeRate не было параметра даты, я его добавлял, поэтому date из datetime затенялась. Изменил параметр ExchangeRate на on_date, вернул импорт обратно.

idlesign replied:
Ааа, вот ты о чём. Атрибуты класса вполне могут называться также как же, как встроенные типы, например.

То есть, в записи
```
class ExchangeRate(NamedTuple):
    id: str
    date: date
    name: str
    ...
```
ничего смутительного нет. Мы тут не перекрываем ни аннотацию, ни определение атрибута, ведь доступаться до него мы будет через `obj.date`. Если бы находились, например в инициализаторе, который принимал бы аргумент с названием `date`, тогда, да, внутри метода инициализации затенялось бы определение типа `date`.

idlesign commented on tests/test_rates.py:
```diff
@@ -1,15 +1,140 @@
+
+    with pytest.raises(WrongArguments) as e:
+        lib[None]
+    assert e.value.message == 'You must pass ISO code, numeric code or code the Bank of Russia of currency' \
```
По форматированию. Башни кода в правой части и пустота в левой читаются не очень.
Поэтому здесь для облегчения лучше делать:
```
assert e.value.message == (
    'You must pass ISO code, numeric code or code the Bank of Russia of currency' 
    'or Currency instance. Not None.'
)
```

А вообще, здесь и в подобных случаях ниже, если текст исключения узнаваем по его части, то полностью его вносить в тесты смысла нет.

idlesign commented on tests/test_rates.py:
```diff
@@ -23,3 +148,161 @@ def test_rates():
     assert rates['840'].name == 'Доллар США'
+
+
+def test_exchange_rates_extra():
```
Нужно подумать, возможно стоит использовать вместо сквозных тестов (с реальными походами в ЦБ) имитаторы (фикстура response_mock).

suhanoves replied:
Думаю это можно вынести отдельным Issue, я потом это изучу и попробую реализовать

idlesign replied:
Хорошо.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+        return hash((self.id, self.num, self.code))
+
+    def __eq__(self, cls):
+        return isinstance(cls, type(self)) and (cls.id, cls.num, cls.code) == (self.id, self.num, self.code)
```
Тут и выше. А зачем так сложно? Кажется необходимым и достаточным сличить лишь какой-то из идентификаторов, например id или code.

suhanoves replied:
Решил хешировать по трём, чтобы уменьшить вероятность замещения какой-либо валюты при деноминации, если какой-то из кодов решат не изменять. Ну и сравнение тогда уж проводить тоже по трём.
Думаешь это сильно влияет на время исполнения?

idlesign replied:
Время исполнения — копейки, тут скорее речь про усложнение излишнее.
При взгляде непосвященного на такой, создаётся впечатление, что это всё действительно важно.
Коды просто так, к частью, не меняют, ни в ISO, ни в ЦБ %)

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+            date of loading the latest information from www.cbr.ru
+        currencies (dict of Currency):
+            a {Union[Currency.id, Currency.num, Currency.code]: Currency} dictionary for all currencies
+    """
```
Форматирование: если описание многострочное, то пустая строка перед закрывающими """ (PEP 257).

А атрибуты можно разметить прямо в коде:
```
TypeCurrecyIdent = Union[Currency.id, Currency.num, Currency.code]
...
self.currencies: Dict[TypeCurrecyIdent, Currency] = None
"""dictionary for all currencies"""
```

suhanoves replied:
> Форматирование: если описание многострочное, то пустая строка перед закрывающими """ (PEP 257).

Перечитал PEP несколько раз, не увидел данной строчки, поэтому не очень понимаю, что ты имел ввиду.
Зато заметил, что docstring класса всегда должен отделяться пустой строкой после, поправил где надо было.

suhanoves replied:
> TypeCurrecyIdent = Union[Currency.id, Currency.num, Currency.code]

так не получится, ведь Union принимает типы, а не значения. Наверное надо просто написать словами, что это словарь, где ключами выступают "id, num or code of currency"

idlesign replied:
> Перечитал PEP несколько раз, не увидел данной строчки, поэтому не очень понимаю, что ты имел ввиду.

Похоже, я застрял в 2014 году. Как я стар %)
https://github.com/python/peps/commit/ab85e78f9b2a06c8c2b7866e0577787f21346fd5#diff-17aa10ddb5326f762412f43f6ffdae8bb9fcf118f91c7d05cb11e20ec41fdc63

> так не получится, ведь Union принимает типы, а не значения. 

Да уж, жаль. Вот эта чепуха `Dict[TypeCurrecyIdent, Currency]`, по смыслу должна сводится к `Dict[str, Currency]`, а о то, что там ид, код, номер можно действительно словами дописать.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+    def __getitem__(self, value: Union[int, str]) -> Optional[Currency]:
+        """Returns Currency by dictionary lookup, converting the argument to ISO format."""
+        if not value:
+            raise WrongArguments(f"You must pass ISO code, numeric code or code the Bank of Russia of currency or "
```
Форматирование: чтобы не было башен кода с правой стороны зачастую достаточно переносить строку сразу после открывающей скобки.
Вообще по большому счёту тут проверкой наверное можно пренебречь вовсе. Гипотетически сюда конечно могут передать любой объект, оценивающийся как «пустой», но практически мы всё равно подымем `CurrencyNotExists`. Поэтому в таком случае лучше в описание для `CurrencyNotExists` передать код, запрошенный пользователем: 
```
riase CurrencyNotFound(f'Currency "{value}" is not found.')
```

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+        if not value:
+            raise WrongArguments(f"You must pass ISO code, numeric code or code the Bank of Russia of currency or "
+                                 f"Currency instance. Not {'None' if value is None else 'empty string'}.")
+        item_ = self._format_num_code(value)
```
А зачем подчерк в конце?

suhanoves replied:
до value у меня функция принимала item :)

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+
+        return currency
+
+    def update(self):
```
По смыслу это не `update`, а `reset` или `init`.

suhanoves replied:
Функция перезаписывает значения, но если мы добавили что вручную, она остаётся, пока существует экземпляр, поэтому reset точно не подходит.
А чем update не нравится? Мы получаем новые сведения, если есть старые, затираем поверх, если старых нет, добавляем.

Вопрос остаётся открытым. Как скажешь, так и назову.

idlesign replied:
> Функция перезаписывает значения, но если мы добавили что вручную, она остаётся, пока существует экземпляр, поэтому reset точно не подходит.

Повторный вызов update полностью перетирает то, что накоплено при помощи add, ведь `self.currencies = self._parse(raw_data)`.

suhanoves replied:
верно, `self.currencies = self._parse(raw_data)`, но внутри `_parse` мы проходимся по байтовой строке, распознавая валюты, и применяем к `self.currencies` метод `update`, тем самым не затирая старые валюты, но обновляя, + добавляем новые если появились.

idlesign replied:
Дэ к это означает, что ответственность за несвязанные по сути действия размазана по двум методам.

suhanoves replied:
> Дэ к это означает, что ответственность за несвязанные по сути действия размазана по двум методам.

Не очень понял, что ты имеешь ввиду? Возвращать при помощи метода parse только словарь, а для обновления метода использовать self.currencies.update(parsed_data)?

idlesign replied:
Кажется тут мы говорим о разных вещах. В `Currencies._parse()` разбираются данные и они заменяют `self.currencies` в `.update()`.

> и применяем к self.currencies метод update,

Вот этого я не вижу в `_parse()`, я вижу создание нового словаря `currencies = {}`.
Но если бы я это увидел, то спросил бы: почему ответственность за обновление лежит на двух методах, включая `_parse()`.

suhanoves replied:
Да, ты прав. Это я затупил, я там обновляю локальный словарь, и потом его присваиваю `self.currencies`.
Предлагаю тогда:
1. Переношу всё, что связано с библиотекой в отдельный модуль, там сохраняю в коде константу-словарь с валютами.
2. В  `__init__` Currencies инициировать `self.currencies` из словаря сохранённого в коде
2. В методе `update()` переписать self.currencies.update(self._parse(raw_data))

И да, спасибо за терпение)

idlesign replied:
Можно так, ага.

> 2. В __init__ Currencies инициировать self.currencies из словаря сохранённого в коде

Если это будет словарь, то там особо инициализировать нечего. Простой импорт уже будет его инициализацией.

> И да, спасибо за терпение)

Тебе спасибо за инициативу.

suhanoves replied:
Игорь, я сейчас думаю над реализацией этого решения, и стою на развилке. Currencies -- это ведь словарь, где ключ - строка с каким либо кодов валюты, а значение - экземпляр валюты. Как тогда захаркорджить это в коде?
1. Самый простой вариант к которому я пока склоняюсь - сохранить в коде две константы, равной строкам xml, которые мы получаем с сайта ЦБ РФ, а при инициализации распарсить их.
2. Создать tuple из экземпляров валют, прописанных ручками (`Currency('id'='R01010', ..., 'code'='AUD'), а в ините просто создать из него уже словарь, где разобрать по кодам. Так хоть константа будет короче в коде, меньше строк.
3. Создавать сразу полноценный словарь вида:
```{'R01010': (aud := Currency('id'='R01010', ..., 'code'='AUD')), 'aud': aud, '036': aud, ...)```
Но его длина 150 строчек текста)

idlesign replied:
Да, тут возможны варианты. Разберём парочку:

1. Положить xml, похожий на ЦБшный (но включающий все коды) в дистрибутив и его разобрать при первом обращении к Currencies в нужный вид.
2. Использовать промежуточное представление данных в виде словаря, максимально приближенного к xml от ЦБ. При инициализации модуля, где хранится этот словарь, перегнать его в нужное представление так, чтобы это представление подменило изначальный объект словаря. При инициализации Curencies импортируем наш модуль со словарём и делаем словарь из него атрибутом объекта Curencies.

Даже из описания видно, что второй вариант более затратен в реализации и не позволяет повторно использовать тот же _parse. Если мы оставляем механизмы разбора из xml, то первый вариант выглядит более выигрышным.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
 
 
-class ExchangeRates(WithRequests):
+class CurrenciesLib(WithRequests, FormatMixin, metaclass=SingletonMeta):
```
Дифф шумит, показывая, что здесь заменён `ExchangeRates`. Это неудобно. Чтобы такого не было, новый класс можно попробовать перенести в конец файла.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+    @staticmethod
+    def _make_currency_pack(currency: Currency) -> Dict[str, Currency]:
+        """Creates a dict of three elements from the currency with the keys: id, code and num if they exists"""
+        pack = dict()
```
Литералом — меньше писать: `{}`.

suhanoves replied:
У меня на литерал ругается Pycharm, он не любит литералы, когда следующей строчкой идёт добавление в словарь.
![30 08 2021 15:53:42](https://user-images.githubusercontent.com/67975624/131342164-14550166-b6d4-4f33-ae9f-3bfbcdebb087.png)
Заменил.

idlesign replied:
> У меня на литерал ругается Pycharm, он не любит литералы, когда следующей строчкой идёт добавление в словарь.

Если известны базовые элементы словаря, то словарь целесообразно прямо с ними инициализировать
```
pack = {
    currency.id.lower(): currency
}
```

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+    def __init__(self):
+        self.update_date = None
+        self.currencies = None
+        self.update()
```
Эту штуку лучше сделать ленивой, потому как пользователям (особенно корпоративным) могут не понравится эти походы в сеть, когда не требуется (каждый дополнительный запрос увеличивает вероятность на то, что он завершится ошибкой). 

Кажется будет правильным, что когда человек использует `ExchageRates`, то подгрузка должна осуществляться при обращении к `currency` курса (). То есть, атрибуты курса, которые можно взять из выдачи оттуда должны продолжать браться, как сейчас.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+                    # ISO numeric code like '036' is loaded like '36', so it needs format to ISO 4217,
+                    # also data from the Bank of Russia contains replaced currencies that do not have ISO attributes.
+                    # additional If-statement was added to exclude format None
+                    num=self._format_num_code(_) if (_ := props['ISO_Num_Code']) else None,
```
Нужно помнить, что для обратной совместимости требуется, чтобы адресовать валюты можно было как прежде (без нулей).

suhanoves replied:
В смысле запрашивать? Запрашивать же можно как угодно сейчас, хоть integer.

В твоей версии и не было проблемы с ведущим нулём. т.к. в запросе курсов он всегда стоит, и валюта будет '036'.
Но в библиотеке валют она может быть проставлена как просто '36', поэтому приведена к ISO.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+        return currencies
+
+    def __str__(self):
+        return f"CurrenciesLib. Update {dt.datetime.strftime(self.update_date, '%Y-%m-%d')}"
```
Тут можно не форматировать, потому что время может тоже быть интересно.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+
+    @property
+    def id(self):
+        return self.currency.id
```
Выше был комментарий: тут, наверное всё же со свойствами не получится, чтобы загрузка была ленивой.

suhanoves replied:
Как вариант задел на версию 2.0, если обратную совместимость убрать, то лениво можно сделать.

idlesign replied:
Так ведь нам сейчас для обратной совместимости важно заполнить id name code num как они раньше заполнялись, без доп. запросов к выдаче списка валют.

suhanoves replied:
Мне тут идея в голову пришла, а что если просто инициализировать библиотеку из файла? Мне кажется это решает все вопросы несколькими килобайтами.

Во-первых, мы не лезем в интернет для создания библиотеки.
Во-вторых, если требуется актуальная версия библиотеки - пользователь применяет метод update (про незатирание я выше ответил)
В-третьих, можно добавить метод dump_data, который этот файл на диске будет обновлять. И при обновлении версии pycbrf его применять, тем самым поставляя с приложением  более менее актуальную библиотеку.
В-четвёртых, не надо придумывать как делать загрузку ленивой и полностью сохраняется старый функционал. Бритва Оккама, прям )

Как ты на это смотришь?

idlesign replied:
> Мне тут идея в голову пришла, а что если просто инициализировать библиотеку из файла? 

Угу, я об этом упоминал ранее. Ты предлагаешь к этому добавить примочку в виде обновления. Но, как было сказано ранее, такие штуки меняются относительно редко, поэтому примочка по большей части будет не востребована.

suhanoves replied:
В итоге мне как переписать код?
Я со своей стороны вижу три выхода:
1. Сделать все модули (библиотеку валют, запрос курса валюты по дате, запрос диапазона курсов) вообще не связанными между собой. Т.е. в курсы валюты подставляется не валюта, а просто заполняются аргументы из запроса. А библиотека так просто рядом будет, чтобы к ней обращаться справочно.
2. Оставить экземпляр валюты в курсах. Этот экземпляр валюты берётся из библиотеки, которая инициализируется из файла, но имеет методы обновления себя из запроса, дампа файла  текущей библиотеки и метод обновить+сбросить дамп за 1 раз.
3. Вариант, который ты отметаешь, тот что пока сделан сейчас. При запуске модуля отправлять запрос на получение библиотеки.

idlesign replied:
Я немного раньше описывал, как это могло бы работать так, чтобы было обратно совместимо: заполняем всё как сейчас, а свойство currency позволяет добраться до дополнительной информации и при доступе к нему в первый раз инициализируется Currencies. И подход с хранением справочника в коде (скажем в словаре), тоже неплох. А вот к гибриду (хранение + обновление не через выпуск новой версии) сразу возникает вопрос: где хранить файл с данными так, чтобы его можно был обновлять (был доступ на записи в файловую систему). Такие вопросы лучше не пытаться решать, если их можно обойти.

Поэтому, если есть возможность сделать полный дамп и загнать в код, то можно так и сделать. Но ручное обновление Currencies практически теряет смысл, так как никто не будет ситуационно дёргать библиотеку — будут либо дёргать обновление всегда (что равноценно варианту, описанному мною вначале), либо никогда. Это, конечно, если она сама не будет пытаться подтягивать данные, когда их недостаёт.

suhanoves replied:
> свойство currency позволяет добраться до дополнительной информации и при доступе к нему в первый раз инициализируется Currencies

Игорь, у меня нет опыта разработки за плечами, поэтому я вижу здесь решение этой задачи через написание своих дескрипторов. Есть более  простое решение?

idlesign replied:
Есть уже готовые штуки для конструирования дескрипторов, в частности `@property`.

Например, что-то типа:
```
@property
def currency(self):
    return Currency(self.code)
```

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+    """Gets and stores exchange rates for different currencies for a selected date
+
+    Attributes:
+        date_requested (date):
```
Атрибуты лучше описать по их месту инициализации (см. пример выше).

suhanoves replied:
Я решил раскидать аттрибуты написать в классе, а в __init__ описывать аргументы. PEP это позволяет. На мой взгляд так логичнее, что ли. Смотря докстринг класса ты видишь все аттрибуты, к которым можешь обращаться, а в __init__ ты видешь именно то, что должен передать.
Но если скажешь перенести - перенесу

idlesign replied:
Сейчас мало кто и редко читает код полностью. Обычный сценарий: либо открыть документацию (а в ней описания к атрибутам видны), либо использовать навигацию в ide (структурную, где видны все атрибуты, или контекстную, в раскрывающемся списке; например в pycharm ctrl+q на атрибуте поднимается как раз документацию к атрибуту).

Но если очень хочется, конечно, можно продублировать и в описании к классу.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+
+    def __init__(self, on_date: Union[str, dt.date, dt.datetime, None] = None, locale_en: bool = False):
+        """
+        Args:
```
У меня в проектах использует форматирование строк документации в rst формате, а тут похоже на что-то гуглове.

suhanoves replied:
Переделаю всё под rst, но я с ним знаком плохо, проверь меня тут дополнительно, пожалуйста.

idlesign replied:
Можно в настройках pycharm - tools - python integrated tools - docstring format выставить rst.
Тогда, если под определением функции напечатать тройные кавычки и нажать enter, параметры окажутся в строке документации автоматом.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,138 +1,508 @@
+    def __init__(self, on_date: Union[str, dt.date, dt.datetime, None] = None, locale_en: bool = False):
+        """
+        Args:
+            on_date (str, date, datetime, None):
```
Типы описаны прямо в сигратуре, дополнительно указывать их в документации не требуется и даже вредно (с точки зрения возможной дальнейшей актуализации).

suhanoves replied:
в примерах документации Sphinx везде пишут :type param:
Убирать?

idlesign replied:
Типы расписывать нет нужды: в sphinx есть модуль, который в документацию подтянет описания типов из объявлений.

suhanoves replied:
Убрал

idlesign commented:
Дифф сейчас слишком сильно шумит. Я частично посмотрел, но ты перенеси новые сущности в конец файла, иначе изменения показываются не там, где они были.

suhanoves commented:
> Дифф сейчас слишком сильно шумит. Я частично посмотрел, но ты перенеси новые сущности в конец файла, иначе изменения показываются не там, где они были.

Currency сдвинуть вниз не получится, т.к. она используется в объявлении ExchangeRate.
Currencies, ExchangeRateDynamics я менял местами, вставлял между, всё равно diff получается шумным, т.к. PyCharm цепляется не за те методы.
Как вариант можно разнести по разным модулям.

suhanoves committed:
Correcting issues that occurred during a pull request

1. All documentation is converted to the Sphinx docstring format
2. CurrenciesLib renamed to Currencies
3. Banks.py converted to version from commit ad068241
4. Exception CurrencyNotExists renamed to CurrencyNotFound
5. Exception ExchangeRateNotExists renamed to ExchangeRateNotFound
6. Rewritten the _format_num_code method to `f'{val}'.zfill(3)`
7. Long strings aligned with guidelines to the recommendations.
8. In the Currency class, __getitem__ method, the exceptions with an empty attribute are replaced with CurrencyNotFound.
9. In the Currency class, __getitem__ method, the `item_` variable has been renamed to `item`.
10. In the Currency class, _make_currency_pack method, the variable `pack` is declared as a literal {}
11. In the Currency class, __str__ method, the date formatting has been changed to datetime

suhanoves commented:
Я постарался привести файлы чтобы diff лучше увидел изменения, но это не сильно помогло для rates.py (все остальные файлы стало сильно проще сравнивать с коммитом ad068241)

Ченжлог:
1. Вся документация приведена к формату Sphinx docstring (по мануалам [тут](https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html#the-sphinx-docstring-format) и [тут](https://thomas-cokelaer.info/tutorials/sphinx/docstring_python.html))
2. CurrenciesLib переименован в Currencies
3. banks.py приведён к версии из коммита ad068241
4. Исключение CurrencyNotExists переименовано в CurrencyNotFound
5. Исключение ExchangeRateNotExists переименовано в ExchangeRateNotFound
6. Переписан метод _format_num_code на `f'{val}'.zfill(3)`
7. Висящие строчки приведены в соответствие с рекомендациями
8. В классе Currencies в методе __getitem__ заменены исключения при пустом аттрибуте на CurrencyNotFound
9. В классе Currencies в методе __getitem__ переменная item_ переименована в item
10. В классе Currencies в методе _make_currency_pack переменная pack объявлена литералом {}
11. В классе Currencies в методе __str__ убрано форматирование у времени обновления библиотеки.

На незакрытые вопросы жду ответов для дальнейшего приведения кода к требованиям.

idlesign commented:
Что-то я закрутился, прошу прощения. На неделе постараюсь поглядеть и отписаться.

И ремарка: длинные описания обычно не пишут — и время берегут, и потому что часто видно из кода, что и как происходит. Но при обзоре могут, конечно, какие-то уточняющие вопросы появиться.

idlesign commented:
> Как вариант можно разнести по разным модулям.

Да, давай так попробуем. Потому, если что я пакет перенесём всё по курсам.

suhanoves committed:
Currencies init from constants, rename dt.date to date, remove types from docstrings

suhanoves commented:
Т.к. библиотека Currencies сейчас инициализируется из константы, встаёт вопрос, стоит ли делать инициализацию ленивую загрузку?

Сейчас каждый вид курсов имеет аттрибут currencies_lib, который ведёт на единственный экземпляр библиотеки. Это дёшево, т.к. экземпляр один, и просто решает вопрос с добавлением валюты из курсов, которой нет в библиотке.

Не вижу с текущей реализацией необходимости в какой-то дополнителной ленивой загрузке, ведь больше в интернет мы не лезем. Что думаешь?

idlesign commented:
> Не вижу с текущей реализацией необходимости в какой-то дополнителной ленивой загрузке, ведь больше в интернет мы не лезем. Что думаешь?

Да, возможно,  только непонятно, почему атрибут  `currencies_lib` в объектах курсов.  Например, можно рассмотреть вариант инициализации его на уровне модуля и присвоения атрибуту модуля прямо в `currencies.py`:

```
CURRENCIES = Currencies()
```

Только там ещё по коду есть недочёты с оформлением и с читаемостью дифа. Доразнеси, пожалуйста, по модулям: все новые сущности — в новые модули. Тогда можно будет ещё поглядеть с на сложность кода и на протечки абстракций. Возможно есть шанс снизить общую появившуюся сложность.

suhanoves committed:
Splitting the code into separate modules. Moving Currencies from attributes to global namespaces

suhanoves commented:
По максимуму разбил на отдельные модули, чтобы не пересекались новые и старые классы, но git всё равно шумит в rates.py на классе `ExchangeRates`, особвенно в методе `__init__`.
Я как мог подвигал строчки кода, чтобы они соответствовали твоему, удалял и вновь добавлял docstring'и, но большего достичь не смог. Если будут вопросы - задавай

idlesign commented on pycbrf/constants.py:
```diff
@@ -0,0 +1,5 @@
+
+DAILY_CURRENCIES = b'<?xml version="1.0" encoding="windows-1251"?><Valuta name="Foreign Currency Market Lib"><Item ID="R01010"><Name>\xc0\xe2\xf1\xf2\xf0\xe0\xeb\xe8\xe9\xf1\xea\xe8\xe9 \xe4\xee\xeb\xeb\xe0\xf0</Name><EngName>Australian Dollar</EngName><Nominal>1</Nominal><ParentCode>R01010    </ParentCode><ISO_Num_Code>36</ISO_Num_Code><ISO_Char_Code>AUD</ISO_Char_Code></Item><Item ID="R01015">< ...
+
+MONTHLY_CURRENCIES = b'<?xml version="1.0" encoding="windows-1251"?><Valuta name="Foreign Currency Market Lib"><Item ID="R01025"><Name>\xc0\xeb\xe1\xe0\xed\xf1\xea\xe8\xe9 \xeb\xe5\xea</Name><EngName>Albanian Lek</EngName><Nominal>100</Nominal><ParentCode>R01025    </ParentCode><ISO_Num_Code>8</ISO_Num_Code><ISO_Char_Code>ALL</ISO_Char_Code></Item><Item ID="R01030"><Name>\xc0\xeb\xe6\xe8\xf0\xf1\ ...
```
Это очень похоже на DAILY_CURRENCIES, данные дублируются. Мы можем это сделать единой переменной?

suhanoves replied:
Можем, только есть ли в этом практический смысл, ведь тянутся они с двух разных страниц: [этой](https://www.cbr.ru/scripts/XML_valFull.asp?d=0) и [этой](https://www.cbr.ru/scripts/XML_valFull.asp?d=1)?
Странно, если данные дублируются, там вроде разные валюты.

Если когда-нибудь придётся обновлять, придётся совершать дополнительные действия, а это чистые байтовые строки, что получаем при помощи get с центробанка.

Сообщи, что думаешь в итоге, я переделаю если надо (нужно будет чуть `_parse` тогда переписать, он сейчас принимает обе сроки в кортеже)

idlesign replied:
Я понял, давай оставим так.

idlesign commented on pycbrf/currencies.py:
```diff
@@ -0,0 +1,154 @@
+        return hash((self.id, self.num, self.code))
+
+    def __eq__(self, cls):
+        return isinstance(cls, type(self)) and (cls.id, cls.num, cls.code) == (self.id, self.num, self.code)
```
issubclass()?

suhanoves replied:
Я в какой-то книге читал, что `isinstance` используется для проверки того, является ли объект экземпляром класса, а `issubclass` используется для проверки, является ли класс подклассом другого класса. Поэтому так и привык проверять, чтобы код показывал, что именно я хочу проверить.

Не мог бы ты пояснить, почему считаешь, что `issubclass` здесь подходит лучше?

suhanoves replied:
Так же почитал документацию, поэкспериментировал в idle и получил `issubclass() arg 1 must be a class`, а мы ведь сравниваем объект. Придётся точно так же использовать преобразование, только другого аргумента :)

Наверное я тут неправильно обозвал аттрибут, надо `cls` поменять на `obj`

idlesign replied:
> Наверное я тут неправильно обозвал аттрибут, надо cls поменять на obj

Угу, сбивает %)

idlesign commented on tests/test_rates.py:
```diff
@@ -23,3 +29,107 @@ def test_rates():
+    assert rates['MDL'].rate == Decimal('4.19277')
+
+
+# def test_raising_exceptions():
```
Лучше не оставлять закомментированный код — людям потом неясно, для чего он закомментировать и что с ним делать. В большинстве случаев он либо лежит, как ненужный для конца времён, либо его выпиливают не разбираясь.

idlesign commented on pycbrf/rates.py:
```diff
@@ -1,103 +1,136 @@
         """
-        Flag indicating whether rates were returned on exact same date as requested.
-        Note that Bank of Russia won't issue different rates for every day of weekend.
+        self.length = 0
```
А это нам для чего?

suhanoves replied:
Пережитки старых мыслей, когда хотел хранить по другому. Но сейчас просто берём длину словаря.
Удалил, спасибо!

idlesign commented on pycbrf/rate_dynamics.py:
```diff
@@ -0,0 +1,164 @@
+            If currency and one date are passed, get the rate of the specified currency at the date.
+            If only currency is passed, get the exchange rate for today.
+        """
+        self.length = 0
```
И это зачем?

suhanoves replied:
Аналогично верхнему. Удалил, спасибо что заметил

idlesign commented:
Я добавил ещё пару комментариев. Разберём их, да я волью, потом сам подрихтую остальное, а то мы затянули это дело.

> Я как мог подвигал строчки кода, чтобы они соответствовали твоему, удалял и вновь добавлял docstring'и, но большего достичь не смог. 

Угу, похоже из-за строки документации класса, в которой описаны параметры из инициализатора. Давай уберём всё же. Можно минимальный пример использования оставить, но в целом примерам место в документации или readme.

suhanoves commented:
Если что меня не беспокоит количество итераций, я на этом только учусь :) но понимаю, что для тебя это лишний напряг.
Дай тогда ответ по поводу констант, надо ли сливать в одну, если надо я перепишу метод и солью. Остальное подправил.

idlesign commented:
> но понимаю, что для тебя это лишний напряг.

Не то чтобы напряг, скорее потому что описывать текстом дольше, чем подправить %)

suhanoves committed:
Rename cls to obj in __eq__ for Currency

suhanoves committed:
Remove unnecessary length attribute for ExchangeRateDynamics

suhanoves committed:
Change the docstring of ExchangeRates for smooth git diff

suhanoves committed:
Remove unnecessary tests for throwing exceptions

suhanoves commented:
Залил все последние исправления по замечаниям.
Ты когда внесёшь исправления в коде после.перед мёрджем, маякни как-нибудь, пожалуйста.
Я, во-первых, хочу поучиться и сделать для себя выводы. А во-вторых реализовать получение курсов металлов, как просили в соседней теме :)

idlesign commented:
Спасибо, влил.
Можно по истории поглядеть, какие я изменения внёс. 
Если будут вопросы по ним, можно прямо тут задать.
